src/UserSpaceInstrumentation/AllocateInTracee.cpp

```diff
@@ -62,6 +60,21 @@ namespace {
     return ErrorMessage(absl::StrFormat("Failed to read from tracee's memory: %s",
                                         backup_or_error.error().message()));
   }
+  const std::vector<uint8_t> backup = std::move(backup_or_error.value());
+
+  orbit_base::unique_resource restore_memory_and_registers_on_return{
+      pid, [start_address, backup, &original_registers](pid_t pid) {
```

Comment:
Of course this is kind of hacky because the resource is `pid` but the deleter carries state. We could add some "resourceless"-`unique_resource`-like class to `OrbitBase`, or add an ad-hoc class here. Let me know what you think.

Reply:
That's a good question. I've been thinking about this use-case myself before. I dont' see the problem with the deleter being stateful. That is explicitly support. You may have state in there! This also applies for unique_ptr.

My biggest concern would be readability and semantics because you made `pid` a resource but the resource does not really goes out of scope here. It's just a trick. And any other parameter in the capture list could be also considered to be part of the resource.

That being said: I think it's a concern but a rather small one.

Nevertheless, let's look at some alternatives:
1. What I have seen in the past in code-bases is to misuse shared_ptr for this. Also not great in terms of readability but it doesn't have the semantic confusion: `std::shared_ptr<void> cleanup{nullptr, deleter_lambda};`
2. Some code bases have a function called `scope_exit` or `finally` that take a function object and return an opaque object. The function object gets called when the object gets out of scope. I think what you also already had in mind. We could add something to the code base but that would probably be out of scope for this PR. (Some code bases also call that function `defer` because there is something similar in Go which is called `defer`. But since it works on functions and not on scope I would rather make distinction here.

Reply:
> My biggest concern would be readability and semantics because you made pid a resource but the resource does not really goes out of scope here. It's just a trick. And any other parameter in the capture list could be also considered to be part of the resource.

Yep, that's what I meant, I wrote it horribly. So we are on the same page!

Regarding the rest, point 2 is what I had in mind.
I had thought about point 1 but it looked hacky. But it's probably less hacky than using the `unique_resource`. I have a question though: Does that not work with `unique_ptr`?

So I'm leaning towards changing to point 1 for simplicity, I was using `unique_resource` only because it's used in other parts of the module, but the lambda carries no state there so at least there is only a single "resource" (the pid). But let me again know if you have a preference.

Reply:
AFAIK you can't create a `unique_ptr<void>`